agent/xds/listeners.go

```diff
@@ -0,0 +1,229 @@
+		return nil, errors.New("nil config given")
+	}
+
+	// One listener for each upstream plus the public one
+	resources := make([]proto.Message, len(cfgSnap.Proxy.Upstreams)+1)
+
+	// Configure public listener
+	var err error
+	resources[0], err = makePublicListener(cfgSnap, token)
+	if err != nil {
+		return nil, err
+	}
+	for i, u := range cfgSnap.Proxy.Upstreams {
+		resources[i+1], err = makeUpstreamListener(&u)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return resources, nil
+}
+
+// makeListener returns a listener with name and bind details set. Filters must
+// be added before it's useful.
+//
+// Note on names: Envoy listeners attempt gracefull transitions of connections
+// when their config changes but that means they can't have their bind address
+// or port changed in a running instance. Since our users might choose to change
+// a bind address or port for the public or upstream listeners, we need to
+// encode those into the unique name for the listener such that if the user
+// changes them, we actually create a whole new listener on the new address and
+// port. Envoy should take care of closing the old one once it sees it's no
+// longer in the config.
+func makeListener(name, addr string, port int) *v2.Listener {
+	return &v2.Listener{
+		Name:    fmt.Sprintf("%s:%s:%d", name, addr, port),
+		Address: makeAddress(addr, port),
+	}
+}
+
+// makeListenerFromUserConfig returns the listener config decoded from an
+// arbitrary proto3 json format string or an error if it's invalid.
+//
+// For now we only support embedding in JSON strings because of the hcl parsing
+// pain (see config.go comment above call to patchSliceOfMaps). Until we
+// refactor config parser a _lot_ user's opaque config that contains arrays will
+// be mangled. We could actually fix that up in mapstructure which knows the
+// type of the target so could resolve the slices to singletons unambiguously
+// and it would work for us here... but we still have the problem that the
+// config would render incorrectly in general in our HTTP API responses so we
+// really need to fix it "properly".
+//
+// When we do that we can support just nesting the config directly into the
+// JSON/hcl naturally but this is a stop-gap that gets us an escape hatch
+// immediately. It's also probably not a bad thing to support long-term since
+// any config generated by other systems will likely be in canonical protobuf
+// from rather than our slight variant in JSON/hcl.
+func makeListenerFromUserConfig(configJSON string) (proto.Message, error) {
+	// Try to decode directly as a types.Any and return that (with nested types
+	// already proto serialized).
+	var any types.Any
+	err := jsonpb.UnmarshalString(configJSON, &any)
+	if err == nil {
+		return &any, nil
+	}
+	// Failed, try again decoding as a straight listener (i.e. no @type annotation
+	// in the JSON)
+	var l v2.Listener
+	err = jsonpb.UnmarshalString(configJSON, &l)
+	if err != nil {
+		return nil, err
```

Comment:
is there value in retaining the first error from the `jsonpb` unmarshal failure and returning that here too in a multierror?